enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,513 @@
+
+Q: Does CCM provide an API? How is it hosted? HTTPS endpoint? Through load balancer?
+A: No. The CCM vendors part of the code serving cloud features from the [cloud-provider](https://github.com/kubernetes/cloud-provider). CCM provides a secure TCP connection on port [10258](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/cloud-provider/ports.go#L22) and it only provides metrics and readiness/health endpoint.
+
+Q: What happens if the kcm leader has no access to ccm? Can it continue its work? Will it give up leadership?
+A: KCM and CCM perform independently after the initial migration process. KCM will be unaffected by CCM being down after the initial migration. The CCM cluster operator will be responsible for transitioning to Unavailable or Degraded in a similar situation.
+
+Q: What are the thoughts about certificate management?
+A: At present, only the GCP provider has any CSR approval mechanism built in. This functionality is used to approve `NodeClient` and `NodeServer` certificates for Nodes on GCP.
+This functionality should not be required as OpenShift handles certificate approval via the `cluster-machine-approver` component.
+
+Q: Does every node need a CCM?
+A: No. The cluster only needs one active `CCM` at any time. A `Deployment` will manage the `CCM` pod and will have 2 replicas which will use leader election to nominate an active leader. [Source](https://kubernetes.io/docs/concepts/overview/components/#cloud-controller-manager)
+
+Q: How metrics are affected by the CCM migration?
+A: Currently the `CCM` implementation is using the same set of metrics from the core repository, all existing metrics will still be available post migration. The `CCMO` will be required to create a `ServiceMonitor` to instruct prometheus to collect metrics from the `CCM` pod.
+
+## Other notes
+
+### Removing a deprecated feature
+
+Upstream is currently planning to remove support for the in-tree providers with kubernetes 1.23. The enhancement [describes](https://github.com/kubernetes/enhancements/blob/473d6a094f07e399079d1ce4698d6e52ddcc1567/keps/sig-cloud-provider/20190125-removing-in-tree-providers.md) steps towards this process. Our goal is to have a working alternative for the in-tree providers ready to be used on-par with the old implementation before upstream remove this feature from core Kubernetes repository.
+
+The `cloud-provider` and `cloud-config` flags from `kubelet` will be removed in 1.23 - [link](https://github.com/kubernetes/kubernetes/pull/90408).
+
+## Timeline
+
+### Upstream
+
+Follow the Kubernetes community discussions and implementation/releases
+
+- vSphere support has graduated to beta in [v1.18](https://github.com/kubernetes/enhancements/issues/670) (Openshift 4.5)
+
+- Azure support goes into beta in [v1.20](https://github.com/kubernetes/enhancements/issues/667) (Openshift 4.7)
+
+### OpenShift
+
+#### OCP 4.7
+
+- OpenStack platform on out-of-tree `CCM` graduates as tech-preview.
```

Comment:
hmm, are you sure this is feasible? 

To be tech preview in 4.7, we need at least tech preview support for OpenStack CSI migration; and the Storage Team doesn't have anything planned for 4.7, except a sketch of a few (bad) ideas: https://github.com/bertinatto/enhancements-1/commit/af844363666331e9124573bf26f55da1ac0699ac

Reply:
This patch should allow us to enable csi migration for testing: https://github.com/openshift/installer/pull/4304

Reply:
I think the idea here, @Fedosin can confirm, was that we would have a very big warning that says that migration is not included and existing volumes will not be migrated. The intention being that users would only be using this feature on temporary testing clusters and not production systems where storage is a requirement.

We should expand this section to include this detail